cpp/src/arrow/compute/kernels/vector_rank.cc

```diff
@@ -63,21 +63,112 @@ void MarkDuplicates(const NullPartitionResult& sorted, ValueSelector&& value_sel
+    };
+    MarkDuplicates(sorted, value_selector);
+  }
+  return sorted;
+}
+
+template <typename ArrowType>
+Result<NullPartitionResult> DoSortAndMarkDuplicate(
+    ExecContext* ctx, uint64_t* indices_begin, uint64_t* indices_end,
+    const ChunkedArray& input, const std::shared_ptr<DataType>& physical_type,
+    const SortOrder order, const NullPlacement null_placement, bool needs_duplicates) {
+  auto physical_chunks = GetPhysicalChunks(input, physical_type);
+  if (physical_chunks.empty()) {
+    return NullPartitionResult{};
+  }
+  ARROW_ASSIGN_OR_RAISE(auto sorted,
+                        SortChunkedArray(ctx, indices_begin, indices_end, physical_type,
+                                         physical_chunks, order, null_placement));
+  if (needs_duplicates) {
+    const auto arrays = GetArrayPointers(physical_chunks);
+    auto value_selector = [resolver = ChunkedArrayResolver(span(arrays))](int64_t index) {
+      return resolver.Resolve(index).Value<ArrowType>();
+    };
+    MarkDuplicates(sorted, value_selector);
+  }
+  return sorted;
+}
+
+template <typename InputType>
+class SortAndMarkDuplicate : public TypeVisitor {
+ public:
+  SortAndMarkDuplicate(ExecContext* ctx, uint64_t* indices_begin, uint64_t* indices_end,
+                       const InputType& input, const SortOrder order,
+                       const NullPlacement null_placement, const bool needs_duplicate)
+      : TypeVisitor(),
+        ctx_(ctx),
+        indices_begin_(indices_begin),
+        indices_end_(indices_end),
+        input_(input),
+        order_(order),
+        null_placement_(null_placement),
+        needs_duplicates_(needs_duplicate),
+        physical_type_(GetPhysicalType(input.type())) {}
+
+  Result<NullPartitionResult> Run() {
+    RETURN_NOT_OK(physical_type_->Accept(this));
+    return std::move(sorted_);
+  }
+
+#define VISIT(TYPE)                                                                 \
+  Status Visit(const TYPE& type) {                                                  \
+    ARROW_ASSIGN_OR_RAISE(                                                          \
+        sorted_, DoSortAndMarkDuplicate<TYPE>(ctx_, indices_begin_, indices_end_,   \
+                                              input_, physical_type_, order_,       \
+                                              null_placement_, needs_duplicates_)); \
+    return Status::OK();                                                            \
+  }
+
+  VISIT_SORTABLE_PHYSICAL_TYPES(VISIT)
+
+#undef VISIT
 
-  Result<Datum> CreateRankings(ExecContext* ctx,
-                               const NullPartitionResult& sorted) override {
+ private:
+  ExecContext* ctx_;
+  uint64_t* indices_begin_;
+  uint64_t* indices_end_;
+  const InputType& input_;
+  const SortOrder order_;
+  const NullPlacement null_placement_;
+  const bool needs_duplicates_;
+  const std::shared_ptr<DataType> physical_type_;
+  NullPartitionResult sorted_{};
+};
+
+struct PercentileRanker {
```

Comment:
Can we keep the comment here?
```suggestion
// A helper class that emits rankings for the "rank_percentile" function
struct PercentileRanker {
```

Reply:
Done.